### ISSUE TITLE
[Bug]: "Rescan folders" takes many hours when using script

### ISSUE THREAD

**Issue by steve1977** [RandomNinjaAtk/docker-lidarr-extended#132]
There is a closed report with similar / same issue (https://github.com/RandomNinjaAtk/docker-lidarr-extended/issues/70). Re-opening here as I am not sure whether closed reports are being read. @Bait-Fish

The issue is that somehow a "full rescan" is triggered, which takes many hours due to the size of the library.

The issue is replicable and seems to be related to the script. Let me share when it happens and also when it does not happen:

1) Start lidarr-extended docker
* It works for a few hours
* Then "rescan" kicks in and it takes hours to complete

2) Restart docker
* Script again works for a bit
* After a few hours, the rescan kicks in, which takes many hours

3) Restart docker again
* Manually start "rescan folder" and it complete in 2 minutes
* Script then works for 24 hours until the next scan kicks in

I also noticed that "refresh artists" also sometimes kicks in when using the script, which may or may not be related.

**Comment by steve1977** [comment]
One thought: It may be related to "refresh artists". If I run "rescan folder" before I run "rescan folder", it completes within minutes. If I run "refresh artist" and then "rescan folder", then "rescan folders" takes ages.

When I turn off the script, the issue goes away. It seems that the script may mark all artists as "changed" or something like this?

**Comment by Bait-Fish** [comment]
This has continued, and worsened for me. The last rescan took >24 hours, "a day ago"  so could keep good track. And the library hasn't changed much. It's been getting longer and longer. Just as the Audio script indexing has. 

I'll look into you steps and see if I have the same behavior. My experience is the rescan happens every 24 hours. I've gotten used to watching for it, then restating the container. Generally that would end it for the remaining day. Then, wait for the Audio script to reindex for 2.5 hours before it starts pulling down again.

**Comment by RandomNinjaAtk** [comment]
@Bait-Fish I'm going to work on changing the process for what you call re-indexing to speed it up.

Honestly for the scan issue, I have no idea what would be causing it. My only thought would be to disable all scripts in the Lidarr's Settings -> Connection page and see if that improves it.

**Comment by RandomNinjaAtk** [comment]
@Bait-Fish - Give the latest update (1.0.289) a try, this should fix the indexing issue. Hasn't been tested much at all though....

**Comment by Bait-Fish** [comment]
Updated and running 289 now. Interesting. Looks like it may grab missing albums as it processes? I like. I'll give it time to get some data back to you.

What is the preferred paste site, pastebin? Been a while . . .

And, my media is not pristine. it has grabs from other sources. Somehow I accidentally had some media get dumped into the Artist folder and is not subfoldered. But this is not a new problem, old files. I'm manually deleting as I have time. But this has not sped anything up.

Maybe I can peer into the data and see what these 56,000 items are. I just don't know how to find that. Possibly then I can target a problem, like similar repeats or some errata. It seems whenever that number would change, it has happened a few times quickly, unexpectedly bulking by a few thousand. Maybe then I could altogether remove a problem artist?

And, I did not get to try any of the suggestions by @steve1977, opting to go with your update first for feedback.

Thanks as always.

**Comment by RandomNinjaAtk** [comment]
Pastebin should be fine.

**Comment by Bait-Fish** [comment]
Audio 289 log so far. Seems to work fine. Much more satisfying than waiting hours for any progress.
https://pastebin.com/PENPRbGM

Audio 282 recent scan time as an example of what was.
https://pastebin.com/4PJwmGg4

Another random observation, unrelated. In the last week or two, the importing scripts at start seems to take longer. LE webui seems to take longer as a consequence. Not sure why it would take 3 minutes, as of this morning.

**Comment by RandomNinjaAtk** [comment]
What do you mean by importing scripts?

**Comment by Bait-Fish** [comment]
Before Audio and Video scripts are give the 2 minute wait, this shows, 

11/20/2022 8:43:30 AM
Removing previous scripts...
11/20/2022 8:43:30 AM
Importing extended scripts...
11/20/2022 8:46:14 AM
Complete...

**Comment by RandomNinjaAtk** [comment]
Its the permissions fix, that is causing that:
https://github.com/RandomNinjaAtk/docker-lidarr-extended/blob/6f716420a6092fd8fa07fdf9a88466ceb53c8be4/root/etc/cont-init.d/98-script-setup.bash#L31-L36

**Comment by Bait-Fish** [comment]
Not sure if 289 will address the daily rescan issue. I'll keep it going to see.

Posting right now to update on the speed of 289. It is much slower with this method. At the moment, my progress is,

11/20/2022 10:23:36 PM
2022-11-20 22:23:36 :: Audio :: 1.0.289 :: Downloading page 829... (20700 - 20725 of 56999 Results)

And that's from starting around 9 am today. Prior to 289 it would have been done in 6 hours or less. As much as I like seeing results come in sooner instead of waiting 2+ hours, the overall time delay/extension now does not seem worth it.

But that isn't what we're fixing. I'll keep watching for a change in daily rescan time and let this thing run.

**Comment by RandomNinjaAtk** [comment]
@Bait-Fish update to the latest, tried to speed it up a bit more... honestly, it can only go so fast...

**Comment by Bait-Fish** [comment]
I may have spotted something odd. As the indexing count ratchets up, at some point the count reverts back to #3. See timestamp 2022-11-21 15:33:53.

I had to truncate the middle of the file where artists 1-42 were processed. Pastebin complained about size. 

https://pastebin.com/9DJbWWew

**Comment by RandomNinjaAtk** [comment]
Odd, just from memory, I don’t think there is anything that would or could cause that based on the design. But I’ll take another look at it tomorrow.

**Comment by Bait-Fish** [comment]
The counter reversion also happened on 289. It had counted up to 1150 (timestamp 2022-11-21 03:09:59), processed 10 items, then went back to 3 (2022-11-21 03:20:17).

https://pastebin.com/8EJaFnQr

**Comment by RandomNinjaAtk** [comment]
That is expected behavior

**Comment by RandomNinjaAtk** [comment]
@Bait-Fish - You need to update to latest, I will not troubleshoot old code.

**Comment by Bait-Fish** [comment]
I have not observed the long rescan issue since. All I have done is remove ended/deceased artists and updated as dev recommends. A caveat, LE has only run about 28 hours each time and may not be hitting the rescan time.

**Comment by steve1977** [comment]
Same here, it still has endless rescan for me.

One more insight for troubleshooting. I disabled both audio & video script (running usenet instead). And the issue with the long rescan is still there.

**Comment by RandomNinjaAtk** [comment]
Again, you need to disable the post processing scripts. I have a feeling that is what is causing it. But honestly if you want to use them, and if that is the case. There is nothing to be done about it….

**Comment by steve1977** [comment]
There are only two scripts there?

MetadataPostProcess.bash
PlexNotify.bash

If I disable them, no more tagging takes place?

**Comment by RandomNinjaAtk** [comment]
Lidarr mainly does the tagging, assuming you have it enabled

**Comment by steve1977** [comment]
So, this script takes care of Replaygain and embeds lyrics?

Or you referring to the audio-pp.bash post-processing done by sabnzb-extended?

**Comment by RandomNinjaAtk** [comment]
This is unrelated to any other containers.

**Comment by versatilist** [comment]
Can confirm this issue. While updates are nice, this means that the script seems to run constantly (on startup, and nightly scheduled). Takes 22:09:55 to startup my lidarr now. This must run before the deemix script can run, so it really makes checking any change you make (version updates) or I make (config updates) a non-starter. 

**STATUS :: LIDARR BUSY** 

Perhaps you can fork lidarr and make it possible to disable these _Scheduled Tasks_. I don't see much reason to rescan and touch each file every day anyway.

( I did disable the Connection script on PostProcess but I dont see how thats related to this issue as the script only applies when a file is retagged).

**Comment by RandomNinjaAtk** [comment]
I will not be making any changes to lidarr itself. That is out of scope and beyond my skill set.

The reason for disabling the post processing scripts is because my theory is the scripts are adding files to the final directories (lyrics, cover art) or possibly changing the modify times, so lidarr sees the changes and detects them, which causes a rescan as a result. So not having any post processing scripts should mean that only lidarr itself touched the files and prevent it.

But again, this is just a theory and you’d have to run it for a while to see if that is really the case.

**Comment by Bait-Fish** [comment]
I'm trying a new LE container (Unraid) that has nothing changed from defaults, audio/video scripts disabled, and only post processing(just now) removed. I have it scanning the same library and will see in a few days how it behaves. I'll report back then.

**Comment by versatilist** [comment]
With AMD the configuration could be updated, you could make changes to the script and push (frequent) changes as deezer changed, and there was also a separate logfile (terminal window) to track progress. I could also kill the docker or manually stop just that portion. While lidarr was running, stable, and without a restart. 

Most importantly the foreground Task in Lidarr did not interrupt the AMD script. 

Now, while it _appears_ more cohesive, it actually works less reliably. As each time Lidarr is restarted it needs an extensive cleanup process, and your add-ons are interrupted by these Tasks. 

I think I will check out the [steroids ](https://hub.docker.com/r/youegraillot/lidarr-on-steroids/) approach where deemix exists parallel to trackers / indexers and lets lidarr manage it as one of three options to grab a download.  

Hopefully you can reconsider, and maybe even bring back AMD. If you are not into making changes to Lidarr I think this merged approach may be flawed.

**Comment by steve1977** [comment]
Let's wait for the results of @Bait-Fish's results. I actually love how it has been integrated. Now running sabnzbd/sonarr/radarr all in the extended version and like the additional benefits it brings.

I would not even mind not using the audioscript, but the videoscript is quite *gold*... Unfortunately, even when disabling both audio & video scripts and just downloading via usenet leads to this "bug". It is odd though as it runs for quite some hours really well (including occasional short folder rescans). But after some time, it "hangs" as we noticed. I don't think we have yet identified the root cause of the issue.

**Comment by Bait-Fish** [comment]
The test container finished it's first scan in just over 2 days time. I'm only peeking in on its status every once in a while. It's already rescanning, 6 hour so far. I'll keep monitoring without any other Lidarr instances involved. Maybe it will settle out and shorten after a couple days. 

I have a thought to next test this library against plain Lidarr as well.

**Comment by versatilist** [comment]
I also notice that there are a number of pending tasks in the Queue (below _Scheduled_ on **Tasks**). Mostly Rescan Folders.

It is currently running a scheduled Rescan Folders from **13 days ago**. 

I manually removed two dozen scheduled tasks, including the one added on startup.

I also notice that this run, the 137k files are running through the  ImportDescisionMaker much faster. Reading at 50MB/s.

- Last duration: 21:50:14
- This duration: (pending)
- Version 1.1.0.2649

For anyone with this issue, I strongly suggest you manually press the 'x' next to each scheduled task to clear out.

**Comment by RandomNinjaAtk** [comment]
Video is not dependent on Lidarr for imports, that's why there is no pausing for that script. The audio script pauses to keep everything running smoothly in terms of imports and processing by Lidarr.

**Comment by steve1977** [comment]
Unfortunately, the issue goes beyond pausing the script though. During the endless rescan, all importing is also stopped (including importing files downloaded via usenet).

Let's see what @Bait-Fish's trial with the stock lidarr result in. Maybe something was changed in the mainline that triggered this issue independent of the script. It is odd that this issue even exists when disabling the scripts.

**Comment by Bait-Fish** [comment]
So far extended is not at fault in my opinion. This should be tested against a plain Lidarr install as well, which I have not yet done. [hoping someone else helps with that test, I will be delayed]

This current test, a fresh container, and default LE that never had audio/video scripts enabled, and has post processing disabled, is also running very long folder rescans. It has not yet stopped, 19 hours and continuing on this last daily run.

To recap this test, The "first start" folder rescan took >48 hours. The following day, 6 hours and almost 15 minutes. And this current one, 19 hours and counting.

The reason I quoted "first start" above is, I first set this container up as a pure extended install a month prior. I let it run for two days and gave up. So that is one wrinkle in this test. After cancelling, I returned to using my regular LE container. So, for the last test, the first start >48 hour run was a restart/continuation from the true first start which I interrupted in frustration.

I think I'll run this another day, maybe two. I ask that someone else with this similar problem try their library against a plain Lidarr (not LE) install. I have some needed Unraid software/hardware updating and rebooting to do. I'll get that going once this test is done. In the interest of progress, more parallel testing would speed us up.

**Comment by versatilist** [comment]
After restart, the deleted tasks [returned](https://imgur.com/a/NB3DFN9) to the Queue. It is again running the _Rescan Folders_ from 14 days ago. 

Stopped again at 11761/12774. Especially odd since there are 48565 albums in the db. 

Perhaps there was db corruption with a version around this time. 

What is the best procedure to migrate a vanilla lidarr db into extended?

**Comment by RandomNinjaAtk** [comment]
If you want to start clean, stopping the container, purging the `/config` directory, then restart the container and it'll be like you just installed it.

**Comment by versatilist** [comment]
I did a fresh install, keeping only the Lidarr.db from vanilla. 

This seems to have improved the issue, less frequent Rescan Folders. When run, only about 14mins for ~8TB. Still runs ~hourly instead of Nightly. 

DB size (vanilla): 1GB (13GB covers)
DB size (extended; corrupted/rescan bug): 1.2GB (13GB covers)
DB size (extended; clear /config, keep db): 1.4GB (19GB covers)

Took about 10hrs and running for Refresh Artist because all the cover art. I noticed a ton of cleanup like:
- RefreshAlbumService
- MediaCoverService
- DiskScanService
- TrackGroupingService
- DownloadService

It also seemed to find a few albums that vanilla Lidarr did not previously import properly.
- ImportDecisionMaker
- IdentificationService
- ImportApprovedTracks

**Comment by versatilist** [comment]
I found the reason for the DB size discrepancy; my default Standard profile does not include singles. Once removed and refreshed, the size is back as before.

I found that if I simply restart the docker after the Rescan Folders task has task has started, it will not resume nor start again until next scheduled. Then the RNA amd script will countdown for liftoff.

**Comment by Bait-Fish** [comment]
This may be a way to delay the daily rescan, to something like once every 7 days (10080 seconds). This link mentioned how to in Sonarr. https://www.reddit.com/r/sonarrv3/comments/fhtvsc/change_sonarr_refresh_times/ but the names are different in Lidarr.

In Lidarr's appdata, lidarr.db can be opened with a SQLite browser, such as "sqlitebrowser" by linuxserver's Repo on Unraid.

So in my LE's appdata, lidarr.db, I can select ScheduledTasks, then Browse Data. Here I see these TypeNames (manually typed with cold hands may have spelling errors)

NzbDrone.Core.Download.RefreshMonitoredDownloadsCommand (Interval 1)
NzbDrone.Core.Messaging.Commands.MessagingCleanupCommand (Interval 5)
NzbDrone.Core.Update.Commands.ApplicationUpdateCheckCommand (Interval 360)
NzbDrone.Core.HealthCheck.CheckHealthCommand (Interval 360)
NzbDrone.Core.Music.Commands.RefreshArtistCommand (Interval 1440)
NzbDrone.Core.MediaFiles.Commands.RescanFoldersCommand (Interval 1440)
NzbDrone.Core.Housekeeping.HousekeepingCommand (Interval 1440)
NzbDrone.Core.Backup.BackupCommand (Interval 10080)
NzbDrone.Core.ImportLists.ImportListSyncCommand (Interval 1440)
NzbDrone.Core.Indexers.RssSyncCommand (Interval 15)

So, I'm guessing, change NzbDrone.Core.MediaFiles.Commands.RescanFoldersCommand interval 10080 seconds for once every 7 days. -yes, applied (interval now "7 days")

The link above also changed RefreshMonitoredDownloadsCommand to 10 (from 1 in my LE default). Maybe that too? -yes, applied (interval now "10 minutes")

Any suggestions. I'm trying these right now. I've been having trouble getting past an initial star scan after having my primary LE container off for a couple weeks. I stopped it after almost three days of continuous scanning. *sigh* NOTE: DB size is 3.6 GB.

On applying the changes, I now see System>Tasks>Scheduled>Rescan Folders next execution listed as "in 7 days." I think I'll let this settle, even if it takes three days to finish this current scan to see how it goes.

**Comment by RandomNinjaAtk** [comment]
I prefer not to touch/modify lidarr at all in anyway.

**Comment by versatilist** [comment]
I tried to touch it.  The (db) change did not stick. Looks like both Lidarr and Sonarr (when the codebase was similar) were hard coded to use their defaults regardless of what the sqlite db says. See:

src/NzbDrone.Core/Jobs/TaskManager.cs

>             var defaultTasks = new[]
>                 {
>                     new ScheduledTask { Interval = 1, TypeName = typeof(RefreshMonitoredDownloadsCommand).FullName },
>                     new ScheduledTask { Interval = 5, TypeName = typeof(MessagingCleanupCommand).FullName },
>                     new ScheduledTask { Interval = 6 * 60, TypeName = typeof(ApplicationUpdateCheckCommand).FullName },
>                     new ScheduledTask { Interval = 6 * 60, TypeName = typeof(CheckHealthCommand).FullName },
>                     new ScheduledTask { Interval = 24 * 60, TypeName = typeof(RefreshArtistCommand).FullName },
>                     new ScheduledTask { Interval = 24 * 60, TypeName = typeof(RescanFoldersCommand).FullName },
>                     new ScheduledTask { Interval = 24 * 60, TypeName = typeof(HousekeepingCommand).FullName },

==

I'm back to a borked instance after two weeks. Same issues as before:
-- The Rescan Folders task runs forever
-- Each docker restart, and each scheduled time both add a new instance of the task to the queue
-- Thus, the number of scheduled Rescan Folders is grows larger with each day
-- The 'x' in the queue does not delete old tasks

On a very rare occasion I can delete all the 'x' tasks and restart the docker 

_Please consider a small patch to Lidarr to change the default folder scan to at least weekly, if not monthly._

**Comment by RandomNinjaAtk** [comment]
I’m not a software developer and this project will not be making changes to
Lidarr. If you want lidarr changed, you’ll need to request it in Lidarr support channels or write a PR for it to the upstream Lidarr project. 

Lidarr here should behave no differently than it currently does in the upstream container…

**Comment by versatilist** [comment]
I respect your position. Understandably, neither of us want to debug or change source of Lidarr. 

Please consider either: 
- keeping an option to use amd via a separate container, or, 
- migrate amd to an indexer (so that it can run in parallel with other tasks) 

Unfortunately my issue may be tied to architectural changes with lidarr when moving to v8. In particular, the Unmapped Files of the many albums I have that are not yet in musicbrainz. Hopefully the database change fixes the issue for others in this thread.

**Comment by RandomNinjaAtk** [comment]
Your issue is with lidarr, so having this in a separate container won't help, because the issue is still with lidarr. If you want the Lidarr team to look at the issue, you need to go through the correct support channels and not here....

**Comment by RandomNinjaAtk** [comment]
The conversation here is un-productive, Lidarr is an unmodified application in this container and suggested mitigations would be to disable all optional post-processing scripts. No evidence has been presented to prove the additional scripts in the container are causing the problem. 

If you want changes made to Lidarr itself, you need to use Lidarr upstream support channels to address your particular issue.

Link: https://github.com/Lidarr/Lidarr

**Comment by Bait-Fish** [comment]
So far today with the default you mention checked, rescanning has occurred in 1 hour chunks then finishing. And this has accumulated to 10 hours of rescanning over the last 12 hours. This left only two hours of active script time.

Just made the suggested change and will monitor and see if it helps. Thanks again for your work, and support.

**Comment by RandomNinjaAtk** [comment]
I might be able to come up with a workaround that would only ignore the rescan task to allow the script to continue. All other tasks would still cause it to pause though. Which might be a good compromise.

**Comment by RandomNinjaAtk** [comment]
@Bait-Fish This workaround should help you out! https://github.com/RandomNinjaAtk/docker-lidarr-extended/commit/3f8436185bc6690c36ca7be712aa640c024348b8

Still not ideal, but it should no longer hold you up.... I haven't given the change much testing other than a quick proof of concept outside of the script...

**Comment by Bait-Fish** [comment]
I let the previous LE finish its full Rescan Folders (45 hours total) which delayed using your latest change.

I restarted LE today to update scripts and am observing.

**Comment by RandomNinjaAtk** [comment]
@Bait-Fish It should always ignore the disk scan task with this change.... However, it will pause for any other task that is not a disk scan (the other tasks typically take much less time because there not as intense)... The reason for pausing is to not have imports overwrite each other and allow Lidarr to properly import the data...

**Comment by Bait-Fish** [comment]
Confirmed. The long scan is chugging along, and the script continues working. Great! 

The Lidarr setting you initially mentioned did not improve the issue while testing prior.

**Comment by RandomNinjaAtk** [comment]
Yeah, wasn't sure it would, the changes are just a workaround at this point.... I don't think anyone determined if it was script related...? Assuming all lidarr "connect" scripts are disabled, it shouldn't be possible for it to be caused by the scripts...

**Comment by Bait-Fish** [comment]
Shortly after you mentioned the scripts in this thread, I did so. Disabling them made no effect as far as I could see. Then after this topic closed, I reenabled them using autoconfig. I don't think they caused any negative effect.
